src/OrchardCore/OrchardCore.Modules/ModularTenantContainerMiddleware.cs

```diff
@@ -41,6 +41,7 @@ public class ModularTenantContainerMiddleware
                 var shellContext = _orchardHost.GetOrCreateShellContext(shellSetting);
 
                 var existingRequestServices = httpContext.RequestServices;
+
                 using (var scope = shellContext.EnterServiceScope())
                 {
                     if (!shellContext.IsActivated)
```

Comment:
Here i also check `shellContext.IsActivating()` because `IsActivated` is set to true before all `Activated` events are executed. So, an incoming request could pass through before, but maybe the only important thing is to only wait that `Activating` events end, not `Activated` ones.

Then, just for info i also added to the `ShellContext` a `SemaphoreSlim` property to do the following
in place of `lock (shellContext)` 

    await shellContext.Semaphore.WaitAsync();